engine-types/src/types/balance.rs

```diff
@@ -0,0 +1,107 @@
+use crate::fmt::Formatter;
+use crate::{Add, Display, Div, Mul, Sub, SubAssign};
+use borsh::{BorshDeserialize, BorshSerialize};
+
+pub const ZERO_BALANCE: Balance = Balance::new(0);
+
+#[derive(
+    Default, Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, BorshSerialize, BorshDeserialize,
+)]
+/// Engine `balance` type which wraps an underlying u128.
+pub struct Balance(u128);
+
+impl Display for Balance {
+    fn fmt(&self, f: &mut Formatter<'_>) -> crate::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Balance {
+    /// Constructs a new `Fee` with a given u128 value.
+    pub const fn new(amount: u128) -> Balance {
+        Self(amount)
+    }
+
+    pub fn checked_sub(self, rhs: Self) -> Option<Self> {
+        self.0.checked_sub(rhs.0).map(Self)
+    }
+
+    pub fn checked_add(self, rhs: Self) -> Option<Self> {
+        self.0.checked_add(rhs.0).map(Self)
+    }
+
+    /// Consumes `Fee` and returns the underlying type.
+    pub fn into_u128(self) -> u128 {
+        self.0
+    }
+}
+
+impl Add<Balance> for Balance {
+    type Output = Balance;
+
+    fn add(self, rhs: Balance) -> Self::Output {
+        Balance(self.0 + rhs.0)
+    }
+}
+
+impl Sub<Balance> for Balance {
+    type Output = Balance;
+
+    fn sub(self, rhs: Balance) -> Self::Output {
+        Balance(self.0 - rhs.0)
+    }
+}
+
+impl SubAssign<Balance> for Balance {
+    fn sub_assign(&mut self, rhs: Balance) {
+        *self = *self - rhs;
+    }
+}
+
+impl Mul<Balance> for Balance {
+    type Output = Balance;
+
+    fn mul(self, rhs: Balance) -> Self::Output {
+        Balance(self.0 * rhs.0)
+    }
+}
```

Comment:
Still, an explanation is needed here. Because there are basic entities Add, Sub - in your vision, everything is fine with them. However, the single difference between Mul is execution priority, and its presence for a single entity is not mutually contradictory.

Reply:
If I have $5 and you have $10, how can we combine our money to create $50? We cannot. This is because multiplying two dollar amounts together does not make sense.

If there is a group of 10 friends, each of whom has $5, how can they combine their money together to create $50? They can pool their money to get $5 + $5 + $5 + $5 + $5 + $5 + $5 + $5 + $5 + $5 = 10 * $5 = $50. Note in this example that 10 is a pure number (not a dollar amount), which is why `Mul<u128> for Balance` would make sense as an implementation, it is a short-cut for repeated addition (or in the language of linear algebra, we would say this unit supports [scalar multiplication](https://en.wikipedia.org/wiki/Scalar_multiplication)).

Multiplication (and division) is fundamentally different than addition (and subtraction) when it comes to quantities that have some semantic meaning in the real world. The same is true for units of distance, time, mass, etc. You can add two quantities that have the same units to get a new quantity with those units (3 apples + 2 apples = 5 apples). You cannot multiply two quantities with the same units to get a new quantity with those units (3 apples * 2 apples = ??? what does it mean to multiply apples?).

To learn more about units and algebraic operations on them, check out [dimensional analysis](https://en.wikipedia.org/wiki/Dimensional_analysis).

Reply:
But still correct: `5 apples * 2 = 10 apples`
In our case it's implementation for:
`Mul<u128> for Balance`

Reply:
Actually I removed it.

Reply:
Yes, `Mul<u128> for Balance` would have been fine, as I said in my rant. But I'm also just as happy to see it removed if we never use it anyways.